### PR TITLE
sqlcipher: add 4.3.0

### PR DESCRIPTION
**Pull request by theodelrieu**
Specify library name and version:  **sqlcipher/4.3.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by theodelrieu** [comment]
@danimtb It seems that there is an issue with pthread, is it on the machine?

>checking for library containing pthread_create... none required
checking for library containing pthread_mutexattr_init... none required

File: recipes/sqlcipher/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    def package_unix(self):
+        self.copy("*sqlite3.h", src="install")
+        self.copy("*.so*", dst="lib", src="install", keep_path=False, symlinks=True)
+        self.copy("*.a", dst="lib", src="install", keep_path=False)
+        self.copy("*.lib", dst="lib", src="install", keep_path=False)
+        self.copy("*.dll", dst="bin", src="install", keep_path=False)
+        self.copy("*.dylib", dst="lib", src="install", keep_path=False)
+        self.copy("*LICENSE", dst="licenses", keep_path=False)
+
+    def package_visual(self):
```
Comment (uilianries) [inline recipes/sqlcipher/all/conanfile.py]:
```suggestion
    def _package_visual(self):
```

File: recipes/sqlcipher/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            tools.patch(**patch)
+        if self.settings.os == "Macos":
+            tools.replace_in_file(os.path.join(self._source_subfolder, "configure"), r"-install_name \$rpath/", "-install_name ")
+
+        if self.settings.compiler == "Visual Studio":
+            self._build_visual()
+        else:
+            self._build_autotools()
+
+    def package_unix(self):
```
Comment (uilianries) [inline recipes/sqlcipher/all/conanfile.py]:
```suggestion
    def _package_unix(self):
```

File: recipes/sqlcipher/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        return {"x86": "i686",
+                "armv6": "arm",
+                "armv7": "arm",
+                "armv7hf": "arm",
+                # Hack: config.guess of sqlcipher does not like aarch64
+                "armv8": "armv8",
+                "mips64": "mips64"}.get(str(self.settings.arch),
+                                        str(self.settings.arch))
+
+    def get_configure_args(self, absolute_install_dir):
```
Comment (uilianries) [inline recipes/sqlcipher/all/conanfile.py]:
```suggestion
    def _get_configure_args(self, absolute_install_dir):
```

File: recipes/sqlcipher/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            else:
+                build = None
+            autotools_env.configure(args=configure_args, host=host, build=build, vars=env_vars)
+            if self.settings.os == "Windows":
+                # sqlcipher will create .exe for the build machine, which is Linux...
+                tools.replace_in_file(os.path.join(self.build_folder, self._source_subfolder, "Makefile"), "BEXE = .exe", "BEXE = ")
+            autotools_env.make(args=["install"])
+
+    @property
+    def arch_id_str_compiler(self):
```
Comment (uilianries) [inline recipes/sqlcipher/all/conanfile.py]:
```suggestion
    def _arch_id_str_compiler(self):
```

File: recipes/sqlcipher/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            host = "%s-linux-elf" % self.arch_id_str_compiler
+        elif self.settings.os == "Windows":
+            arch = str(self.settings.arch)
+            if arch == "x86":
+                arch = "i386"
+            host = "%s-pc-mingw32" % arch
+        elif self.settings.os == "iOS":
+            host = "%s-apple-darwin" % self.settings.arch
+
+        configure_args = self.get_configure_args(absolute_install_dir)
```
Comment (uilianries) [inline recipes/sqlcipher/all/conanfile.py]:
```suggestion
        configure_args = self._get_configure_args(absolute_install_dir)
```

**Comment by theodelrieu** [comment]
Ok, Windows is fixed but I don't think I can do anything about the pthread issue...

**Comment by uilianries** [comment]
@theodelrieu maybe you can try https://github.com/conan-io/conan-center-index/pull/695

**Comment by theodelrieu** [comment]
@uilianries Unfortunately, the problem arises only on Linux, and `pthreads4w` seem to only work for Windows... The thing I don't get is how did `openssl` pass the CI here?

Does the `test_package` in `openssl/1.1.1d` use any symbol of `libcrypto.a`?

File: recipes/sqlcipher/all/test_package/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.0)
+
+project(PackageTest C)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
```
Comment (uilianries) [inline recipes/sqlcipher/all/test_package/CMakeLists.txt]:
```suggestion
conan_basic_setup()
```

File: recipes/sqlcipher/all/test_package/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+add_executable(example example.c)
+target_link_libraries(example CONAN_PKG::sqlcipher)
```
Comment (uilianries) [inline recipes/sqlcipher/all/test_package/CMakeLists.txt]:
```suggestion
target_link_libraries(example ${CONAN_LIBS})
```

**Comment by theodelrieu** [comment]
Whaaat? Is this intended? @uilianries

**Comment by uilianries** [comment]
> Whaaat? Is this intended? @uilianries

No, but CONAN_LIBS is different from targets of course. For some reason, the targets is wrong. But the weird thing is, I only was able to reproduce with GCC5. With GCC9, targets works fine.

**Comment by uilianries** [comment]
That's weird, the new errors from Windows build is because some system functions are not implemented, like, the system lib is missing.

For instance, [ConvertFiberToThread](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-convertfibertothread), which is part of kernel32, but as we know, kernel32 is automatically linked. :thinking:

**Comment by theodelrieu** [comment]
Sigh, let the whack-a-mole part begin...

**Comment by theodelrieu** [comment]
@uilianries In the error log there is the following flag: `/NODEFAULTLIB:kernel32.lib`...

This seems fishy

**Comment by theodelrieu** [comment]
Ok it seems the NODEFAULTLIB flags do not have a real purpose, and the ones that do are reversed. Should(?) be fixed now.

Although I am very doubtful about the `CONAN_LIBS` versus `CONAN_PKG`, something is very wrong here. I'll try to reproduce later

**Comment by uilianries** [comment]
@theodelrieu The problem related to target is the lib list order. I can say CONAN_LIBS uses alphabetic order.

**Comment by theodelrieu** [comment]
@uilianries I remember seeing an issue about it (I might have opened it), it's really a big problem!

Anyway, I think there's nothing else to be done for sqlcipher.

File: recipes/sqlcipher/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                "TCLSH_CMD=%s" % self.deps_env_info.TCLSH,
+                ]
+
+        main_target = "dll" if self.options.shared else "sqlcipher.lib"
+
+        if self.settings.compiler.runtime in ["MD", "MDd"]:
+            nmake_flags.append("USE_CRT_DLL=1")
+        if self.settings.build_type == "Debug":
+            nmake_flags.append("DEBUG=2")
+        if platform.release() == "10":
```
Comment (SSE4) [inline recipes/sqlcipher/all/conanfile.py]: does it mean library built on Windows 10 won't be compatible with lower Windows versions?
also, does it mean resulting library will be different depending if it was built on Windows 10 or another Windows Version?

Comment (theodelrieu) [reply recipes/sqlcipher/all/conanfile.py]: Looking at the `Makefile.msc`:

```
# Set this non-0 to compile binaries suitable for the Windows 10 platform.
#
!IFNDEF FOR_WIN10
FOR_WIN10 = 0
!ENDIF

...

# When compiling for the Windows 10 platform, the PLATFORM macro must be set
# to an appropriate value (e.g. x86, x64, arm, arm64, etc).
#
!IF $(FOR_WIN10)!=0
!IFNDEF PLATFORM
!ERROR Using the FOR_WIN10 option requires a value for PLATFORM.
!ENDIF
!ENDIF
```

I'll try to unconditionally enable `FOR_WIN10`, it seems that it's required in order for built binaries to run everywhere.

**Pull request merged**